File: frontend/src/components/benchmark/content/pagerank/implementations/pagerank.js
class PageRank {
    constructor() {
        this.adjacency = new Map();
        this.nodes = new Set();
    }

    addEdge(u, v) {
        if (!this.adjacency.has(u)) {
            this.adjacency.set(u, new Set());
        }
        this.adjacency.get(u).add(v);
        this.nodes.add(u);
        this.nodes.add(v);
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    pagerank(damping = 0.85, maxIter = 100, tol = 1e-6) {
        const pr = {};
        const n = this.nodes.size;

        if (n === 0) return pr;

        for (const node of this.nodes) {
            pr[node] = 1.0 / n;
        }

        for (let iter = 0; iter < maxIter; iter++) {
            const newPr = {};
            let diff = 0.0;

            for (const node of this.nodes) {
                let rank = (1 - damping) / n;

                for (const neighbor of this.nodes) {
                    if (this.adjacency.has(neighbor) && this.adjacency.get(neighbor).has(node)) {
                        const outDegree = this.adjacency.get(neighbor).size;
                        if (outDegree > 0) {
                            rank += damping * (pr[neighbor] / outDegree);
                        }
                    }
                }

                newPr[node] = rank;
                diff += Math.abs(newPr[node] - pr[node]);
            }

            Object.assign(pr, newPr);
            if (diff < tol) break;
        }

        return pr;
    }

    topKNodes(k = 10, damping = 0.85) {
        const pr = this.pagerank(damping);
        return Object.entries(pr)
            .sort((a, b) => b[1] - a[1])
            .slice(0, k)
            .map(([node, score]) => ({ node: parseInt(node), score }));
    }
}

module.exports = PageRank;

File: frontend/src/components/benchmark/content/pagerank/implementations/pagerank.go
package main

import (
	"math"
	"sort"
)

type PageRank struct {
	adjacency map[int]map[int]bool
	nodes     map[int]bool
}

func NewPageRank() *PageRank {
	return &PageRank{
		adjacency: make(map[int]map[int]bool),
		nodes:     make(map[int]bool),
	}
}

func (pr *PageRank) AddEdge(u, v int) {
	if pr.adjacency[u] == nil {
		pr.adjacency[u] = make(map[int]bool)
	}
	pr.adjacency[u][v] = true
	pr.nodes[u] = true
	pr.nodes[v] = true
}

func (pr *PageRank) BuildFromEdges(edges [][2]int) {
	for _, edge := range edges {
		pr.AddEdge(edge[0], edge[1])
	}
}

func (pr *PageRank) PageRank(damping float64, maxIter int, tol float64) map[int]float64 {
	result := make(map[int]float64)
	n := len(pr.nodes)

	if n == 0 {
		return result
	}

	for node := range pr.nodes {
		result[node] = 1.0 / float64(n)
	}

	for iter := 0; iter < maxIter; iter++ {
		newResult := make(map[int]float64)
		diff := 0.0

		for node := range pr.nodes {
			rank := (1 - damping) / float64(n)

			for neighbor := range pr.nodes {
				if pr.adjacency[neighbor] != nil && pr.adjacency[neighbor][node] {
					outDegree := len(pr.adjacency[neighbor])
					if outDegree > 0 {
						rank += damping * (result[neighbor] / float64(outDegree))
					}
				}
			}

			newResult[node] = rank
			diff += math.Abs(newResult[node] - result[node])
		}

		result = newResult
		if diff < tol {
			break
		}
	}

	return result
}

func (pr *PageRank) TopKNodes(k int, damping float64) [][2]interface{} {
	ranks := pr.PageRank(damping, 100, 1e-6)

	type pair struct {
		node  int
		score float64
	}

	pairs := []pair{}
	for node, score := range ranks {
		pairs = append(pairs, pair{node, score})
	}

	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].score > pairs[j].score
	})

	if k > len(pairs) {
		k = len(pairs)
	}

	result := [][2]interface{}{}
	for i := 0; i < k; i++ {
		result = append(result, [2]interface{}{pairs[i].node, pairs[i].score})
	}

	return result
}